pkg/metrics/metrics.go

```diff
@@ -0,0 +1,250 @@
+// Copyright 2023 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package metrics
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	combine "gonum.org/v1/gonum/stat/combin"
+
+	"github.com/kanisterio/kanister/pkg/log"
+)
+
+// BoundedLabel is a type that represents a label and its associated valid values.
+type BoundedLabel struct {
+	LabelName   string
+	LabelValues []string
+}
+
+// getLabelNames gets the all the `LabelName` fields from each BoundedLabel struct.
+func getLabelNames(bl []BoundedLabel) []string {
+	ln := make([]string, 0)
+	for _, l := range bl {
+		ln = append(ln, l.LabelName)
+	}
+	return ln
+}
+
+// verifyBoundedLabels checks if the BoundedLabel list is valid
+// returns true if valid, and false if invalid.
+func verifyBoundedLabels(bl []BoundedLabel) bool {
+	if len(bl) == 0 {
+		return false
+	}
+	for _, l := range bl {
+		if len(l.LabelValues) == 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// getLabelCombinations takes a slice of BoundedLabel elements and
+// returns a list of permutations of possible label permutations.
+func getLabelCombinations(bl []BoundedLabel) ([]prometheus.Labels, error) {
+	/*
+		Considering the following example - If we have two BoundedLabel elements:
+		BoundedLabel{
+		  LabelName: "operation_type"
+		  LabelValues: ["backup", "restore"]
+		}
+		BoundedLabel{
+		  LabelName: "action_set_resolution"
+		  LabelValues: ["success", "failure"]
+		}
+		The following code generates the permutation list:
+		[ {"operation_type": "backup", "action_set_resolution": "success"},
+		{"operation_type": "backup", "action_set_resolution": "failure"},
+		{"operation_type": "restore", "action_set_resolution": "success"},
+		{"operation_type": "restore", "action_set_resolution": "failure"}]
+	*/
+	if !verifyBoundedLabels(bl) {
+		return nil, errors.New("invalid BoundedLabel list")
+	}
+	resultPrometheusLabels := make([]prometheus.Labels, 0)
```

Comment:
nit: similar to other comments, if we delay the creation of this slice until the permutations have been generated, we can allocate the correctly sized slice:
```suggestion
	resultPrometheusLabels := make([]prometheus.Labels, 0, len(idxPermutations))
```